### PR TITLE
Replace EventBus type with Injection Module event bus strategy

### PR DESCRIPTION
**Pull request by ibaca**
The event bus maps event types so a developer can publish or observe
event of a specific type. EventBus was a quite useful utility, but now
with type injections, module definitions and async libraries like RxJava
you can use the concept of event bus without the explicit event bus. Now
the event bus is implicit in the whole app. If you wan to emmit events,
you inject Observer<EventType> and if you want to listen to events you
inject Observable<EventType>. This IMO is even better than traditional
event bus bc it force you to describe which kind of event you are
actually using.

### CODE REVIEW

**Comment by ibaca** [comment]
You have -830lines of code, same functionality and testability, and IMO and even more explicit/descriptive event usage in your presenters. What do you think? 😃

**Comment by morisil** [comment]
Thanks a lot, sounds like a great simplification. I will take a look ASAP.

File: src/main/java/com/xemantic/githubusers/logic/presenter/UserQueryPresenter.java [hunk 1]
```diff
@@ -35,16 +35,16 @@
  */
 public class UserQueryPresenter {
 
-  private final EventBus eventBus;
+  private final Observer<UserQueryEvent> eventBus;
```
Comment (morisil) [inline src/main/java/com/xemantic/githubusers/logic/presenter/UserQueryPresenter.java]: `userQuerySink` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/DrawerPresenterTest.java [hunk 2]
```diff
@@ -52,7 +50,7 @@
   @Test
   public void start_noInteraction_shouldObserveAllTheIntentsAndDoNothingWithView() {
     // given
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<SnackbarMessageEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/DrawerPresenterTest.java]: `snackbarMessage$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/DrawerPresenterTest.java [hunk 3]
```diff
@@ -78,7 +76,7 @@ public void start_noInteraction_shouldObserveAllTheIntentsAndDoNothingWithView()
   @Test
   public void start_openDrawerIntent_shouldOpenTheDrawer() {
     // given
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<SnackbarMessageEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/DrawerPresenterTest.java]: `snackbarMessage$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/DrawerPresenterTest.java [hunk 4]
```diff
@@ -107,9 +105,9 @@ public void start_openDrawerIntent_shouldOpenTheDrawer() {
   @Test
   public void start_readAboutIntent_shouldPostToSnackbar() {
     // given
-    EventBus eventBus = new DefaultEventBus();
-    EventTracker tracker = new EventTracker(SnackbarMessageEvent.class);
-    tracker.attach(eventBus);
+    PublishSubject<SnackbarMessageEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/DrawerPresenterTest.java]: `snackbarMessage$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/DrawerPresenterTest.java [hunk 5]
```diff
@@ -168,9 +166,9 @@ public void start_openProjectOnGitHubIntent_shouldOpenProjectUrl() {
   @Test
   public void start_selectLanguageIntent_shouldPostToSnackbar() {
     // given
-    EventBus eventBus = new DefaultEventBus();
-    EventTracker tracker = new EventTracker(SnackbarMessageEvent.class);
-    tracker.attach(eventBus);
+    PublishSubject<SnackbarMessageEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/DrawerPresenterTest.java]: `snackbarMessage$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/SnackbarPresenterTest.java [hunk 6]
```diff
@@ -41,7 +40,7 @@
   @Test
   public void start_noEventPosted_shouldDoNothingWithView() {
     // given
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<SnackbarMessageEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/SnackbarPresenterTest.java]: `snackbarMessage$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/SnackbarPresenterTest.java [hunk 7]
```diff
@@ -56,14 +55,14 @@ public void start_noEventPosted_shouldDoNothingWithView() {
   public void start_eventPosted_shouldShowTheMessage() {
     // given
     SnackbarMessageEvent event = new SnackbarMessageEvent("foo");
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<SnackbarMessageEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/SnackbarPresenterTest.java]: `snackbarMessage$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 8]
```diff
@@ -60,7 +59,7 @@
   @Test
   public void start_noInput_shouldDoNothingWithViewAndServices() {
     // given
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<UserQueryEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 9]
```diff
@@ -87,7 +86,7 @@ public void start_noInput_shouldDoNothingWithViewAndServices() {
   @Test
   public void onUserQueryEvent_emptyQueryString_shouldDoNothingWithView() {
     // given
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<UserQueryEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 10]
```diff
@@ -118,7 +117,7 @@ public void onUserQueryEvent_emptyQueryString_shouldDoNothingWithView() {
   public void onUserQueryEvent_1UserToBeFound_shouldRetrieveAndDisplay1UserAndDisableLoadMore() {
     // given
     int totalCount = 1;
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<UserQueryEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 11]
```diff
@@ -218,7 +217,8 @@ public void onLoadMore_when1pageIsAlreadyDisplayedAnd2UsersInTotal_shouldRequest
     // given
     int pageSize = 1;
     int totalCount = 2;
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<UserQueryEvent> userQueryBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 12]
```diff
@@ -218,7 +217,8 @@ public void onLoadMore_when1pageIsAlreadyDisplayedAnd2UsersInTotal_shouldRequest
...
+    PublishSubject<UserSelectedEvent> userSelectedBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userSelection$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 13]
```diff
@@ -275,7 +275,8 @@ public void onLoadMore_when1pageIsAlreadyDisplayedAnd2UsersInTotal_shouldRequest
   public void onLoadAll1000Users_when1001UsersFoundInTotal_shouldRequestAndDisplay1000UsersIn10PagesAndThenDisableLoadMore() {
     // given
     int totalCount = 1001;  // higher than max limit of 1000
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<UserQueryEvent> userQueryBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 14]
```diff
@@ -275,7 +275,8 @@ public void onLoadMore_when1pageIsAlreadyDisplayedAnd2UsersInTotal_shouldRequest
...
+    PublishSubject<UserSelectedEvent> userSelectedBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userSelection$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 15]
```diff
@@ -351,11 +345,11 @@ public void onUserQueryEvent_whenAlreadyHandlingOneQueryAndRequest_shouldCancelO
     given(view.observeLoadMore()).willReturn(loadMoreTrigger);
 
     UserListPresenter presenter = new UserListPresenter(
-        eventBus,
+        events$,
         userService,
         errorAnalyzer,
         () -> mock(UserView.class),
-        () -> new UserPresenter(eventBus),
+        () -> new UserPresenter(PublishSubject.create()),
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: Now I wonder why I just don't mock `UserPresenter` here?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 16]
```diff
@@ -294,18 +295,18 @@ public void onLoadAll1000Users_when1001UsersFoundInTotal_shouldRequestAndDisplay
     given(view.observeLoadMore()).willReturn(loadMoreTrigger);
 
     UserListPresenter presenter = new UserListPresenter(
-        eventBus,
+        userQueryBus,
         userService,
         errorAnalyzer,
         () -> userView,
-        () -> new UserPresenter(eventBus),
+        () -> new UserPresenter(userSelectedBus),
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: now I wonder why I just don't mock `UserPresenter` here?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 17]
```diff
@@ -405,7 +396,7 @@ public void onErrorInRequest_errorIsRecoverable_shouldRecoverWhenLoadMoreIsTrigg
     RuntimeException error = new RuntimeException();
     int totalCount = 2;
     int pageSize = 2;
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<UserQueryEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 18]
```diff
@@ -463,7 +454,7 @@ public void onErrorInRequest_errorIsNonRecoverable_shouldDoNothingOnLoadMoreAndR
     @SuppressWarnings("ThrowableNotThrown")
     RuntimeException error = new RuntimeException();
 
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<UserQueryEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 19]
```diff
@@ -351,11 +345,11 @@ public void onUserQueryEvent_whenAlreadyHandlingOneQueryAndRequest_shouldCancelO
     given(view.observeLoadMore()).willReturn(loadMoreTrigger);
 
     UserListPresenter presenter = new UserListPresenter(
-        eventBus,
+        events$,
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java [hunk 20]
```diff
@@ -531,16 +517,16 @@ public void stop_whileHandlingRequest_shouldUnsubscribeFromEventBusCancelRequest
...
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserListPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserPresenterTest.java [hunk 21]
```diff
@@ -49,7 +47,7 @@
   @Test
   public void start_user_shouldDisplayUser() {
     // given
-    EventBus eventBus = new DefaultEventBus();
+    PublishSubject<UserSelectedEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserPresenterTest.java]: `userSelection$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserPresenterTest.java [hunk 22]
```diff
@@ -71,9 +69,9 @@ public void start_user_shouldDisplayUser() {
   @Test
   public void onUserSelected_view_shouldPostUserSelectedEvent() {
     // given
-    EventBus eventBus = new DefaultEventBus();
-    EventTracker eventTracker = new EventTracker(UserSelectedEvent.class);
-    eventTracker.attach(eventBus);
+    PublishSubject<UserSelectedEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserPresenterTest.java]: `userSelection$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserQueryPresenterTest.java [hunk 23]
```diff
@@ -51,9 +49,9 @@
   @Test
   public void start_view_shouldOnlyBindToView() {
     // given
-    EventBus eventBus = new DefaultEventBus();
-    EventTracker tracker = new EventTracker(UserQueryEvent.class);
-    tracker.attach(eventBus);
+    PublishSubject<UserQueryEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserQueryPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserQueryPresenterTest.java [hunk 24]
```diff
@@ -66,15 +64,15 @@ public void start_view_shouldOnlyBindToView() {
...
+    assertThat(tracker.getOnNextEvents(), empty());
   }
 
   @Test
   public void onUserQuery_queryString_shouldPostEventWithQueryString() {
     // given
-    EventBus eventBus = new DefaultEventBus();
-    EventTracker tracker = new EventTracker(UserQueryEvent.class);
-    tracker.attach(eventBus);
+    PublishSubject<UserQueryEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserQueryPresenterTest.java]: `userQuery$` ?

File: src/test/java/com/xemantic/githubusers/logic/presenter/UserQueryPresenterTest.java [hunk 25]
```diff
@@ -86,16 +84,16 @@ public void onUserQuery_queryString_shouldPostEventWithQueryString() {
...
+    assertThat(tracker.getOnNextEvents().get(0).getQuery(), is("foo"));
   }
 
   @Test
   public void onUserQuery_2subsequentQueriesProvided_shouldPost2EventsWithQueryString() {
     // given
-    EventBus eventBus = new DefaultEventBus();
-    EventTracker tracker = new EventTracker(UserQueryEvent.class);
-    tracker.attach(eventBus);
+    PublishSubject<UserQueryEvent> eventBus = PublishSubject.create();
```
Comment (morisil) [inline src/test/java/com/xemantic/githubusers/logic/presenter/UserQueryPresenterTest.java]: `userQuery$` ?

**Comment by ibaca** [comment]
What about:
```
Observable<TheEvent> theEvent$;
Consumer<TheEvent> theEventConsumer;
PublishSubject<TheEvent> theEventBus;
```
I have changed Observer to Consumer so the user doesn't get confused with complete/error methods which will make the bus die. And, theEventEmtter might be a better name, but theEventConsumer is acceptable and trivially to remember.

File: pom.xml [hunk 26]
```diff
@@ -66,4 +66,22 @@
     </dependency>
   </dependencies>
 
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-source-plugin</artifactId>
```
Comment (ibaca) [inline pom.xml]: uhm... you depends on the source jar but you do not deploy the source jar on install... so it fails. I just add it so mvn install deploy both the jar and the source jar.

**Comment by ibaca** [comment]
I have amended everything. I hope this match your code style. But, IMO this kind of PR it is just easier that you clone my branch, create your commit without asking 😛 as this is your project! and just fix your code style and merge my commit and your adaptation commit together. After that, I think github detect that all the PR commit have been merged, so it closes it automatically. Much faster IMO, this is just an idea with minor code style changes. But, hehe I'm totally ok applying this minor changes myself 👍.

**Comment by morisil** [comment]
I am going for vacation starting from today and will be out of internet for a week. I hope to get back to your PR ASAP.

**Comment by morisil** [comment]
@ibaca thanks for this contribution. It's a great simplification.

**Pull request merged**